### PR TITLE
Enriched refine chance correction

### PR DESCRIPTION
**Pull request by Smokexyz**
### Pull Request Prelude

[//]: # (Thank you for working on improving Hercules!)

[//]: # (Please complete these steps and check the following boxes by putting an `x` inside the brackets _before_ filing your Pull Request.)

- [x] I have followed [proper Hercules code styling][code].
- [x] I have read and understood the [contribution guidelines][cont] before making this PR.
- [x] I am aware that this PR will be closed if the above-mentioned criteria are not fulfilled.

### Changes Proposed

New information from kRO suggests that enriched ore rates are similar to iRO's. Or rather iRO is following kRO's enriched ore success rates table up to +10. 
This PR updates the current refine_db.conf with kRO's enriched refine chances for both pre-re and re. This is also required for the new refinery user interface. Previously, enriched ores would simply re-roll the normal item chance, which is undoubtedly incorrect.

Source: https://www.youtube.com/watch?v=Q1aU69qlYJA&feature=youtu.be&t=89

**Affected Branches:** Master

[cont]: https://github.com/HerculesWS/Hercules/blob/master/CONTRIBUTING.md
[code]: https://github.com/HerculesWS/Hercules/wiki/Coding-Style

### CODE REVIEW

**Comment by HerculesWSAPI** [comment]
This change is [Reviewable](https://reviewable.io/reviews/herculesws/hercules/1684)

**Comment by Asheraf** [comment]
i think this should include the shadow equipments refine as well `npc/re/merchants/shadow_refiner.txt`

**Comment by Jedzkie** [comment]
@Smokexyz maybe you can change also the name of `Chance` to `Normal` because it's kinda annoying to read since you add `Enriched` in the config file.

**Comment by Smokexyz** [comment]
@Jedzkie it is as annoying as any other one-worded config name that is not so easy to understand, it just takes some getting used to. `Enriched` couldn't mean anything else here.

**Comment by dastgir** [comment]
EnrichedChance would be good as it's readable too and self explanatory.

**Comment by Jedzkie** [comment]
@Smokexyz much better. thanks!

**Comment by MishimaHaruna** [comment]
I'm not against this if it is how official servers do it and there is no way to calculate it, but where are the values from? In the video I only see a few of the values (and, as a side note, the few I calculated, appear to be equivalent to the old re-rolling behavior, with the final chance rounded up to the next multiple of 10)

**Comment by Smokexyz** [comment]
@MishimaHaruna They are from iRO wiki - http://irowiki.org/wiki/Refinement_System
Assuming that kRO uses the same values.

**Comment by Smokexyz** [comment]
> (and, as a side note, the few I calculated, appear to be equivalent to the old re-rolling behavior, with the final chance rounded up to the next multiple of 10)

How do you mean? The final rolled chance is not shown, only the threshold.

**Comment by MishimaHaruna** [comment]
Well, I'm just speculating, I don't have actual data to check against (but then again, irowiki is just speculation as well as far as I'm aware)

The values I see in irowiki are very close to the values you obtain if you re-roll once in case of failure:

| Lvl | Refine | iW chance (normal) | iW chance (enriched) | Re-roll calculation |
| ---- | ------- | ---------------------- | ------------------------ | ------------------- |
| 1 | +8 | 60% (0.6) | 90% (0.9) | `1 - (1 - 0.6)^2 = 0.84` (84%) |
| 1 | +9 | 40% (0.4) | 70% (0.7) | `1 - (1 - 0.4)^2 = 0.64` (64%) |
| 1 | +10 | 19% (0.19) | 30% (0.3) | `1 - (1 - 0.19)^2 = 0.3439` (34%) |
| 2 | +7 | 60% (0.6) | 90% (0.9) | `1 - (1 - 0.6)^2 = 0.84` (84%) |
| 2 | +8 | 40% (0.4) | 70% (0.7) | `1 - (1 - 0.4)^2 = 0.64` (64%) |
| 2 | +9 | 20% (0.2) | 40% (0.4) | `1 - (1 - 0.2)^2 = 0.36` (36%) |
| 2 | +10 | 19% (0.19) | 30% (0.3) | `1 - (1 - 0.19)^2 = 0.3439` (34%) |
| 3 | +6 | 60% (0.6) | 90% (0.9) | `1 - (1 - 0.6)^2 = 0.84` (84%) |
| 3 | +7 | 50% (0.5) | 80% (0.8) | `1 - (1 - 0.5)^2 = 0.75` (75%) |
| 3 | +8 | 20% (0.2) | 40% (0.4) | `1 - (1 - 0.2)^2 = 0.36` (36%) |
| 3 | +9 | 20% (0.2) | 40% (0.4) | `1 - (1 - 0.2)^2 = 0.36` (36%) |
| 3 | +10 | 19% (0.19) | 30% (0.3) | `1 - (1 - 0.19)^2 = 0.3439` (34%) |
| 4 | +5 | 60% (0.6) | 90% (0.9) | `1 - (1 - 0.6)^2 = 0.84` (84%) |
| 4 | +6 | 40% (0.4) | 70% (0.7) | `1 - (1 - 0.4)^2 = 0.64` (64%) |
| 4 | +7 | 40% (0.4) | 70% (0.7) | `1 - (1 - 0.4)^2 = 0.64` (64%) |
| 4 | +8 | 20% (0.2) | 40% (0.4) | `1 - (1 - 0.2)^2 = 0.36` (36%) |
| 4 | +9 | 20% (0.2) | 40% (0.4) | `1 - (1 - 0.2)^2 = 0.36` (36%) |
| 4 | +10 | 9% (0.09) | 20% (0.2) | `1 - (1 - 0.09)^2 = 0.1719` (17%) |
| A | +5 | 60% (0.6) | 90% (0.9) | `1 - (1 - 0.6)^2 = 0.84` (84%) |
| A | +6 | 40% (0.4) | 70% (0.7) | `1 - (1 - 0.4)^2 = 0.64` (64%) |
| A | +7 | 40% (0.4) | 70% (0.7) | `1 - (1 - 0.4)^2 = 0.64` (64%) |
| A | +8 | 20% (0.2) | 40% (0.4) | `1 - (1 - 0.2)^2 = 0.36` (36%) |
| A | +9 | 20% (0.2) | 40% (0.4) | `1 - (1 - 0.2)^2 = 0.36` (36%) |
| A | +10 | 9% (0.09) | 20% (0.2) | `1 - (1 - 0.09)^2 = 0.1719` (17%) |

You can see how the values calculated with the re-roll formula are very similar to the enriched rates reported by irowiki (and if you round them to the nearest multiple of 10, they're identical).

Note: the formula I used is the following:

```
  Pr(at least one successful roll)
= 1 - Pr(no successful rolls)
= 1 - Pr(unsuccessful roll) ^ (number of rolls)
= 1 - (1 - Pr(successful roll)) ^ (number of rolls)
```

**Comment by Smokexyz** [comment]
@MishimaHaruna That's really cool. We could probably use this rounded formula instead, if you think it's viable. If yes, It could be hardcoded in `status_get_refine_chance()` and then retrieved where required using this function.

**Comment by Asheraf** [comment]
Just to add some infos here those rates are published by kRO on their Official website http://ro.gnjoy.com/news/probability/View.asp?category=4&seq=1941565&curpage=1
also we already have a topic in the forum about those 
http://herc.ws/board/topic/11527-re-kro-equipment-refine-rate/

**Comment by Smokexyz** [comment]
@Asheraf Thank you, I wasn't aware. I guess then we should specifically use the official rates.

**Comment by MishimaHaruna** [comment]
Ok, thanks for the links @Asheraf, I wasn't aware either. It sounds reasonable to use those rates then.

@Smokexyz could you update the PR to include the remaining values from those tables as well?

**Comment by Smokexyz** [comment]
@MishimaHaruna Done.

**Comment by Asheraf** [comment]
@Smokexyz can you update the documentation for `*getequippercentrefinery` :)

File: db/pre-re/refine_db.conf [hunk 1]
```diff
@@ -39,10 +39,10 @@ Armors/WeaponLevel1~4: { // Specifies weapon level or armor type
 	                                        //   Random bonus start level. This is only applied for weapons, and not displayed client-side.
 	Rates: { // Per level configuration of the refine rates.
 		Lv1~10: { // Level of refine
-			Chance: value     (int, optional, defaults to 100) // Chance of successful refine (100 = 100%)
+			NormalChance: value     (int, optional, defaults to 100) // Chance of successful refine (100 = 100%)
```
Comment (Emistry) [inline db/pre-re/refine_db.conf]: Documentation for these field are missing?
```
EnrichedChance: 
EventNormalChance: 
EventEnrichedChance: 
```

File: db/re/refine_db.conf [hunk 2]
```diff
@@ -39,10 +39,10 @@ Armors/WeaponLevel1~4: { // Specifies weapon level or armor type
 	                                        //   Random bonus start level. This is only applied for weapons, and not displayed client-side.
 	Rates: { // Per level configuration of the refine rates.
 		Lv1~10: { // Level of refine
-			Chance: value     (int, optional, defaults to 100) // Chance of successful refine (100 = 100%)
+			NormalChance: value     (int, optional, defaults to 100) // Chance of successful refine (100 = 100%)
```
Comment (Emistry) [inline db/re/refine_db.conf]: Documentation for these field are missing?
```
EnrichedChance: 
EventNormalChance: 
EventEnrichedChance: 
```

File: npc/merchants/advanced_refiner.txt [hunk 3]
```diff
@@ -123,7 +123,7 @@ S_RefineValidate:
 	mes "Would you like to continue?";
 	next;
 	if(select("Yes", "No") == 1) {
-		if (getequippercentrefinery(getarg(3)) < 100) {
+		if (getequippercentrefinery(getarg(3), true) < 100) {
```
Comment (Emistry) [inline npc/merchants/advanced_refiner.txt]: hmm ... `true`?? aren't suppose to use the documented constant here? like `REFINE_CHANCE_TYPE_XXXXX` ?

File: npc/merchants/advanced_refiner.txt [hunk 4]
```diff
@@ -93,7 +93,7 @@ payon,157,146,6	script	Suhnbi#cash	4_M_03,{
 
 	mes "[Suhnbi]";
 	mes "Clang! Clang! Clang!";
-	if (getequippercentrefinery(.@part) > rand(100) || getequippercentrefinery(.@part) > rand(100)) {
+	if (getequippercentrefinery(.@part, true) > rand(100)) {
```
Comment (Emistry) [inline npc/merchants/advanced_refiner.txt]: hmm ... `true`?? aren't suppose to use the documented constant here? like `REFINE_CHANCE_TYPE_XXXXX` ?

**Comment by Smokexyz** [comment]
@Asheraf @Emistry skipped those bits earlier, updated.

**Comment by MishimaHaruna** [comment]
@Smokexyz looks good to me. Is this ready on your side? (it still has the in-progress flag)

**Comment by Smokexyz** [comment]
@MishimaHaruna I haven't tested this in-game but I have nothing more to add here. So if it looks good please go ahead and merge 👍

**Pull request merged**